aten/src/ATen/core/qualified_name.h

```diff
@@ -1,91 +1,96 @@
+struct QualifiedName {
+  QualifiedName() {}
 
-  // Create a qualified name from a dotted string, splitting it as necessary.
-  // Example input: "foo.bar.baz"
-  static QualifiedNamePtr createFromDotted(const std::string& name) {
+  // `name` can be a dotted string, like "foo.bar.baz", or just a bare name.
+  explicit QualifiedName(std::string name) {
+    AT_ASSERT(!name.empty());
+    // split the string into its atoms.
     size_t startSearchFrom = 0;
-    size_t pos = name.find('.', startSearchFrom);
+    size_t pos = name.find(delimiter_, startSearchFrom);
 
-    auto qualifiedName = QualifiedNamePtr();
     while (pos != std::string::npos) {
       auto atom = name.substr(startSearchFrom, pos - startSearchFrom);
       AT_ASSERTM(
           atom.size() > 0, "Invalid name for qualified name: '", name, "'");
-      qualifiedName = QualifiedName::create(qualifiedName, std::move(atom));
+      atoms_.push_back(std::move(atom));
       startSearchFrom = pos + 1;
-      pos = name.find('.', startSearchFrom);
+      pos = name.find(delimiter_, startSearchFrom);
     }
 
     auto finalAtom = name.substr(startSearchFrom, pos - startSearchFrom);
     AT_ASSERTM(
         finalAtom.size() > 0, "Invalid name for qualified name: '", name, "'");
-    return QualifiedName::create(qualifiedName, std::move(finalAtom));
+    atoms_.push_back(std::move(finalAtom));
+
+    cacheAccessors();
   }
 
-  // Flatten this qualified name and convert the whole thing to a string, like
-  // "foo.bar.baz".
-  std::string toString() const {
-    std::ostringstream ss;
-    toString(ss);
-    return ss.str();
+  // `name` must be a bare name (no dots!)
+  explicit QualifiedName(const QualifiedName& prefix, std::string name) {
+    AT_ASSERT(!name.empty());
+    atoms_.insert(atoms_.begin(), prefix.atoms_.begin(), prefix.atoms_.end());
+    atoms_.push_back(std::move(name));
+
+    cacheAccessors();
   }
 
-  // Compares `this` with `other` for equality, starting from the furthest
-  // prefix.
-  bool equals(const QualifiedNamePtr& other) const {
-    if (!other) {
-      return false;
-    }
+  // The fully qualified name, like "foo.bar.baz"
+  const std::string& qualifiedName() const {
+    return qualifiedName_;
+  }
 
-    if (name_ != other->name_) {
-      return false;
-    }
+  // The leading qualifier, like "foo.bar"
+  const std::string& prefix() const {
+    return prefix_;
+  }
 
-    if (!prefix_) {
-      return !other->prefix_;
-    }
-    return prefix_->equals(other->prefix_);
+  // The base name, like "baz"
+  const std::string& name() const {
+    return name_;
+  }
+
+  bool operator==(const QualifiedName& other) const {
+    return this->qualifiedName_ == other.qualifiedName_;
+  }
+
+  bool operator!=(const QualifiedName& other) const {
+    return !(*this == other);
   }
 
  private:
-  std::ostream& toString(std::ostream& ss) const {
-    if (prefix_ == nullptr) {
-      ss << name_;
-      return ss;
+  char delimiter_ = '.';
```

Comment:
this could be const

Reply:
making members const deletes their copy constructor, which is weird for something that should be a value type. Having it be private and all methods const accomplishes much the same goal